src/scripting.c

```diff
@@ -637,12 +637,62 @@ void scriptingEnableGlobalsProtection(lua_State *lua) {
     sdsfree(code);
 }
 
+/*
+ * parameters:
+ * ud - opaque pointer passed to lua_newstate
+ * ptr - pointer to the block being allocated/reallocated/freed
+ * osize - original size of the block
+ * nsize - the new size of the block
+ * 
+ * NOTE:
+ * ptr is NULL if and only if osize is zero.
+ * When nsize is zero, the allocator must return NULL
+ * if osize is not zero, it should free the block pointed to by ptr.
+ * When nsize is not zero, the allocator returns NULL if and only if it cannot fill the request.
+ * When nsize is not zero and osize is zero, the allocator should behave like malloc.
+ * When nsize and osize are not zero, the allocator behaves like realloc.
+ * 
+ * if server.lua_memory_limit is 0, then the limmitation is disabled
+ */ 
+static void *l_alloc_restricted (void *ud, void *ptr, size_t osize, size_t nsize) {
+    REDIS_NOTUSED(ud);
+    if (ptr == NULL && osize!= 0)
```

Comment:
Remove this check; it's not protecting anything, and the behaviour is not forward compatible.
In lua 5.2+:
> When ptr is NULL, osize encodes the kind of object that Lua is allocating. osize is any of LUA_TSTRING, LUA_TTABLE, LUA_TFUNCTION, LUA_TUSERDATA, or LUA_TTHREAD when (and only when) Lua is creating a new object of that type. When osize is some other value, Lua is allocating memory for something else.